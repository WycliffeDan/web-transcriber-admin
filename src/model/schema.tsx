import { KeyMap, Schema, SchemaSettings } from '@orbit/data';

export const keyMap = new KeyMap();

const schemaDefinition: SchemaSettings =  {
    models: {
        book: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
            },
            relationships: {
              type: { type: 'hasOne', model: 'bookType', inverse: 'books' },
              sets: { type: 'hasMany', model: 'set', inverse: 'book' },
            },
        },
        bookType: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
              description: { type: 'string' },
            },
            relationships: {
              books: { type: 'hasMany', model: 'book', inverse: 'bookType' },
            },
        },
        organization: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
              websiteUrl: { type: 'string' },
              logoUrl: { type: 'string' },
              publicByDefault: { type: 'boolean' },
            },
            relationships: {
              owner: { type: 'hasOne', model: 'user', inverse: 'ownedOrganizations' },
              userMemberships: { type: 'hasMany', model: 'user', inverse: 'organization' },
              users: { type: 'hasMany', model: 'user', inverse: 'organizations' },
              projects: { type: 'hasMany', model: 'project', inverse: 'organization' },
              userRoles: { type: 'hasMany', model: 'userRole', inverse: 'organization' },
            }
        },
        project: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
              dateCreated: { type: 'date' },
              dateUpdated: { type: 'date' },
              dateArchived: { type: 'date' },
              language: { type: 'string' },
              description: { type: 'string' },
              allowDownloads: { type: 'boolean' },
              isPublic: { type: 'boolean' },
              // filter keys
              ownerId: { type: 'number' },
              projectTypeId: { type: 'number' },
              organizationId: { type: 'number' },
            },
            relationships: {
              type: { type: 'hasOne', model: 'projectType', inverse: 'projects' },
              owner: { type: 'hasOne', model: 'user', inverse: 'projects' },
              organization: { type: 'hasOne', model: 'organization', inverse: 'projects' },
              // projectIntegrations: { type: 'hasOne', model: 'projectIntegration', inverse: 'projects' },
              projectUsers: { type: 'hasMany', model: 'user', inverse: 'projects' },
              sets: { type: 'hasMany', model: 'set', inverse: 'project' },
              userTasks: { type: 'hasMany', model: 'userTask', inverse: 'project' },
            }
        },
        projectType: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
              description: { type: 'string' },
            },
            relationships: {
              projects: { type: 'hasMany', model: 'project', inverse: 'type' },
            },
        },
        role: {
            keys: { remoteId: {} },
            attributes: {
              roleName: { type: 'string' },
            },
            relationships: {
              userRoles: { type: 'hasMany', model: 'userRole', inverse: 'role' },
            },
        },
        set: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
            },
            relationships: {
              book: { type: 'hasOne', model: 'book', inverse: 'sets' },
              project: { type: 'hasOne', model: 'project', inverse: 'sets' },
            },
        },
        task: {
            keys: { remoteId: {} },
            attributes: {
              reference: { type: 'string' },
              passage: { type: 'string' },
              position: { type: 'number' },
              taskState: { type: 'string' },
              hold: { type: 'number' },
              title: { type: 'string' },
              dateCreated: { type: 'date' },
              dateUpdated: { type: 'date' },
            },
            relationships: {
              userTasks: { type: 'hasMany', model: 'userTask', inverse: 'task' },
              media: { type: 'hasMany', model: 'taskMedia', inverse: 'task' },
              sets: { type: 'hasMany', model: 'set', inverse: 'task' },
            },
        },
        taskMedia: {
            keys: { remoteId: {} },
            attributes: {
              versionNumber: { type: 'number' },
              artifacttype: { type: 'string' },
              eafUrl: { type: 'string' },
              audioUrl: { type: 'string' },
              duration: { type: 'number' },
              contentType: { type: 'string' },
              audioQuality: { type: 'string' },
              textQuality: { type: 'string' },
              transcription: { type: 'string' },
              dateCreated: { type: 'date' },
              dateUpdated: { type: 'date' },
            },
            relationships: {
              task: { type: 'hasOne', model: 'taskMedia', inverse: 'media' },
            },
        },
        user: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
              givenName: { type: 'string' },
              familyName: { type: 'string' },
              auth0Id: { type: 'string' },
              email: { type: 'string' },
              phone: { type: 'string' },
              isLocked: { type: 'boolean' },
              profileVisibility: { type: 'number' },
              emailNotification: { type: 'boolean' },
              timezone: { type: 'string' },
              localization: { type: 'string' },
              decimalSeparator: { type: 'string' },
            },
            relationships: {
              ownedOrganizations: { type: 'hasMany', model: 'organization', inverse: 'owner' },
              organizationMemberships: { type: 'hasMany', model: 'organization', inverse: 'userMemberships' },
              organizations: { type: 'hasMany', model: 'organization', inverse: 'users' },
              assignedTasks: { type: 'hasMany', model: 'userTask', inverse: 'assigned' },
              projects: { type: 'hasMany', model: 'project', inverse: 'owner' },
              userRoles: { type: 'hasMany', model: 'userRole', inverse: 'user' },
            },
        },
        userRole: {
            keys: { remoteId: {} },
            attributes: {
              roleName: { type: 'string' },
            },
            relationships: {
              user: { type: 'hasOne', model: 'user', inverse: 'userRoles' },
              role: { type: 'hasOne', model: 'role', inverse: 'userRoles' },
              organization: { type: 'hasOne', model: 'organization', inverse: 'userRoles' },
            },
        },
        userTask: {
            keys: { remoteId: {} },
            attributes: {
              activityName: { type: 'string' },
              taskState: { type: 'string' },
              comment: { type: 'string' },
              dateCreated: { type: 'date' },
              dateUpdated: { type: 'date' },
            },
            relationships: {
              project: { type: 'hasOne', model: 'project', inverse: 'userTasks' },
              task: { type: 'hasOne', model: 'task', inverse: 'userTasks' },
              assigned: { type: 'hasOne', model: 'user', inverse: 'assignedTasks' },
            },
        },
    }
};

export const schema = new Schema(schemaDefinition);