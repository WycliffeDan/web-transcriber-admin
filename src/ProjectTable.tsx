import React, { useState, useEffect } from 'react';
import { useGlobal } from 'reactn';
import Project from './model/project';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withData } from 'react-orbitjs';
import { QueryBuilder, Record, TransformBuilder } from '@orbit/data';
import AppBar from '@material-ui/core/AppBar';
import MuiToolbar from '@material-ui/core/Toolbar';
import Paper from '@material-ui/core/Paper';
import Fab from '@material-ui/core/Fab';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import BackIcon from '@material-ui/icons/ArrowBack';
import AddIcon from '@material-ui/icons/Add';
import DeleteIcon from '@material-ui/icons/Delete';
import EditIcon from '@material-ui/icons/Edit';
import IconButton from '@material-ui/core/IconButton';
import { createStyles, withStyles, Theme } from '@material-ui/core/styles';
import { IntegratedSorting, SortingState } from '@devexpress/dx-react-grid';
import { Grid, Table, TableColumnResizing,
  TableHeaderRow, Toolbar } from '@devexpress/dx-react-grid-material-ui';

export class ProjectTableData extends React.Component<IRecordProps, object> {
  public render(): JSX.Element {
      return <ProjectTable {...this.props} />
  }
}

interface Row {
  type: string;
  id: number;
  name: string;
  description: string;
  language: string;
  delete: string;
}

export function ProjectTable(props: any) {
  const { classes, projects, updateStore } = props;
  const [columns, setColumns] = useState([
    { name: 'name', title: 'Name' },
    { name: 'description', title: 'Description' },
    { name: 'language', title: 'Language' },
    { name: 'delete', title: 'Delete' },
  ]);
  const [rows, setRows] = useState([]);
  const [view, setView] = useState('');
  const [project, setProject] = useGlobal('project');

  const handleDelete = (e:any) => {
    updateStore((t: TransformBuilder) => t.removeRecord({
      type: 'project',
      id: e.currentTarget.id
    }))
  };
  const handleAdd = () => { alert('Add') };
  const handleCancel = () => { setView('/admin') };
  const handleEdit = (e:any) => {
    setProject(projects.filter((p: Project) => p.attributes.name.toLowerCase() === e.target.innerText.toLowerCase())[0].id);
    setView('/projectstatus')
  };

  useEffect(() => {
    setRows(projects.map((o: Project) => ({
      type: o.type,
      id: o.id,
      name: o.attributes.name,
      description: o.attributes.description,
      language: o.attributes.language,
      delete: o.id,
    })))
  }, []);

  const LinkCell = ({ value, style, ...restProps }: {value: string, style: object, row: any, column: any, tableRow: any, tableColumn: any}) => (
    <Table.Cell {...restProps} style={{...style}} value >
      <Button
        key={value}
        aria-label={value}
        color="primary"
        className={classes.link}
        onClick={handleEdit}
      >
        {value}
        <EditIcon className={classes.editIcon} />
      </Button>
    </Table.Cell>
  );

  const DeleteCell = ({ value, style, ...restProps }: {value: string, style: object, row: any, column: any, tableRow: any, tableColumn: any}) => (
    <Table.Cell {...restProps} style={{...style}} value >
      <IconButton
        id={value}
        key={value}
        aria-label={value}
        color="default"
        className={classes.deleteIcon}
        onClick={handleDelete}
      >
        <DeleteIcon />
      </IconButton>
    </Table.Cell>
  );

  const Cell = (props: any) => {
    const { column } = props;
    if (column.name === 'name') {
      return <LinkCell {...props} />
    } else if (column.name === 'delete') {
      return <DeleteCell {...props} />
    }
    return <Table.Cell {...props} />
  };

  return view === "" ? (
    <div className={classes.root}>
      <AppBar className={classes.appBar} position="static">
        <MuiToolbar>
          <IconButton >
            <BackIcon onClick={handleCancel} />
          </IconButton>
          <Typography variant="h6" color="inherit" className={classes.grow}>
            {"SIL Transcriber Admin"}
          </Typography>
        </MuiToolbar>
      </AppBar>
      <div className={classes.container}>
        <Paper id="ProjectTable" className={classes.paper}>
        <div className={classes.dialogHeader}>
        <div className={classes.grow} />
        <h2>{"Choose Project"}</h2>
        <div className={classes.grow} />
          <Fab
            key="add"
            aria-label="Add"
            color="primary"
            className={classes.button}
            onClick={handleAdd}
          >
            <AddIcon className={classes.icon} />
          </Fab>
        </div>
        <Grid rows={rows} columns={columns}>
          <SortingState
            defaultSorting={[{ columnName: "name", direction: "asc" }]}
          />
            <IntegratedSorting />
            <Table cellComponent={Cell} />
              <TableColumnResizing
                minColumnWidth={50}
                defaultColumnWidths={[
                  { columnName: "name", width: 200 },
                  { columnName: "description", width: 400 },
                  { columnName: "language", width: 100 },
                  { columnName: "delete", width: 100 }
                ]}
              />
            <TableHeaderRow showSortingControls={true} />
            <Toolbar />
          </Grid>
        </Paper>
      </div>
    </div>
  ) : (
    <Redirect to={view} />
  );
}

const styles = (theme: Theme) => createStyles({
  container: {
    display: 'flex',
    justifyContent: 'center'
  },
  appBar: theme.mixins.gutters({
    background: '#FFE599',
    color: 'black'
  }),
  paper: theme.mixins.gutters({
    paddingTop: 16,
    paddingBottom: 16,
    marginTop: theme.spacing.unit * 3,
    width: '80%',
    display: 'flex',
    flexDirection: 'column',
    alignContent: 'center',
    [theme.breakpoints.down('md')]: {
      width: '100%',
    },
  }),
  grow: {
    flexGrow: 1,
  },
  dialogHeader: theme.mixins.gutters({
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center'
  }),
  editIcon: {
    fontSize: 16,
  },
});

const mapStateToProps = () => ({});
const mapDispatchToProps = (dispatch: any) => ({
    ...bindActionCreators({
    }, dispatch),
});

interface IRecordProps {
    projects: () => Array<Record>;
}

const mapRecordsToProps = {
    projects: (q: QueryBuilder) => q.findRecords('project'),
}

export default withStyles(styles, { withTheme: true })(
    withData(mapRecordsToProps)(
        connect(mapStateToProps, mapDispatchToProps)(ProjectTable) as any
        ) as any
    ) as any;
