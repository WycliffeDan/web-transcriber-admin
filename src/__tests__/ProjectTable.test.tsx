import React from 'react';
import { Router } from 'react-router-dom';
import { setGlobal } from 'reactn';
import { DataProvider } from 'react-orbitjs';
import { Provider } from 'react-redux';
import store from '../store';
import { createMuiTheme } from '@material-ui/core';
import Store from '@orbit/store';
import { Schema } from '@orbit/data';
import { schema, keyMap } from '../schema';
import history from '../history';
import { render, cleanup, waitForElement } from '@testing-library/react';
import 'jest-dom/extend-expect';
import ProjectTable from '../routes/ProjectTable';
import Auth from '../auth/Auth';

const auth = new Auth();

const theme = createMuiTheme({});

const dataStore = new Store({ schema, keyMap });

const globals = {
  organization: null,
  project: null,
  plan: null,
  user: null,
  initials: null,
  lang: 'en',
  dataStore: dataStore,
  schema: schema,
  keyMap: keyMap,
};
setGlobal(globals);

const tree = (props?: any) => (
  <DataProvider dataStore={dataStore}>
    <Provider store={store}>
      <Router history={history}>
        <ProjectTable auth={auth} {...props} />
      </Router>
    </Provider>
  </DataProvider>
);

const addOrgGroupAndProject = async (orgName: string, projName: string) => {
  const organization = {
    type: 'organization',
    attributes: {
      name: orgName,
    },
  } as any;
  (schema as Schema).initializeRecord(organization);
  setGlobal({ ...globals, organization: organization.id });
  await (dataStore as Store).update(t => t.addRecord(organization));
  const group = {
    type: 'group',
    attributes: {
      name: 'Africa',
    },
  } as any;
  (schema as Schema).initializeRecord(group);
  await (dataStore as Store).update(t => t.addRecord(group));
  await (dataStore as Store).update(t =>
    t.replaceRelatedRecord({ type: 'group', id: group.id }, 'owner', {
      type: 'organization',
      id: organization.id,
    })
  );
  const project = {
    type: 'project',
    attributes: {
      name: projName,
    },
  } as any;
  (schema as Schema).initializeRecord(project);
  await (dataStore as Store).update(t => t.addRecord(project));
  await (dataStore as Store).update(t =>
    t.replaceRelatedRecord({ type: 'project', id: project.id }, 'group', {
      type: 'group',
      id: group.id,
    })
  );
};

afterEach(cleanup);

test('can render ProjectTable snapshot', async () => {
  const { getByText, container } = render(tree());
  const TestTable = await waitForElement(() => getByText(/^Description$/i));
  expect(container.firstChild).toMatchSnapshot();
});

test('displays project of selected organization', async () => {
  await addOrgGroupAndProject('SIL', 'Genesis');

  const groups = await (dataStore as Store).query(q => q.findRecords('group'));
  const projects = await (dataStore as Store).query(q =>
    q.findRecords('project')
  );

  const { getByText, container } = render(tree({ projects, groups }));
  const TestTable = await waitForElement(() => getByText(/^Genesis$/i));
  // expect(getByText(/^Genesis$/i)).toHaveTextContent('Genesis');
  const body = container.querySelector('tbody');
  const projectButtons = body && body.querySelectorAll('button');
  projectButtons &&
    projectButtons.forEach(i => {
      if (i.innerText && /^[a-z]/i.test(i.innerText))
        expect(i.innerText).toBe('Genesis');
    });
});

test('displays only projects of selected organization', async () => {
  await addOrgGroupAndProject('SIL', 'Genesis');
  await addOrgGroupAndProject('FCBH', 'Luke');

  const groups = await (dataStore as Store).query(q => q.findRecords('group'));
  const projects = await (dataStore as Store).query(q =>
    q.findRecords('project')
  );

  const { getByText, container } = render(tree({ projects, groups }));
  const TestTable = await waitForElement(() => getByText(/^Luke$/i));
  const body = container.querySelector('tbody');
  const projectButtons = body && body.querySelectorAll('button');
  projectButtons &&
    projectButtons.forEach(i => {
      expect(i.innerText).not.toBe('Genesis');
    });
});
