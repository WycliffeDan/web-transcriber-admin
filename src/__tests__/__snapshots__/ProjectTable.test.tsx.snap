// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render ProjectTable snapshot 1`] = `
<div
  class="WithData-Connect-ProjectTable---root-1"
>
  <header
    class="MuiPaper-root-32 MuiPaper-elevation4-38 MuiAppBar-root-23 MuiAppBar-positionStatic-27 MuiAppBar-colorPrimary-30 Connect-TranscriberBar--appBar-12"
  >
    <div
      class="MuiToolbar-root-59 MuiToolbar-regular-61 MuiToolbar-gutters-60"
    >
      <button
        class="MuiButtonBase-root-69 MuiIconButton-root-63 Connect-TranscriberBar--menuButton-13"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label-68"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-72"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root-245"
        />
      </button>
      <button
        class="MuiButtonBase-root-69 MuiButton-root-81 MuiButton-text-83 MuiButton-flat-86 Connect-TranscriberBar--button-20 WithData-Connect-ProjectTable---button-9"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-82"
        >
          <h6
            class="MuiTypography-root-107 MuiTypography-h6-124 MuiTypography-colorInherit-136"
          />
        </span>
        <span
          class="MuiTouchRipple-root-245"
        />
      </button>
      <h6
        class="MuiTypography-root-107 MuiTypography-h6-124 MuiTypography-colorInherit-136 Connect-TranscriberBar--grow-11 WithData-Connect-ProjectTable---grow-4"
      >
        Transcriber Admin
      </h6>
      <div
        class="Connect-TranscriberBar--grow-11 WithData-Connect-ProjectTable---grow-4"
      />
      <div
        class="Connect-TranscriberBar--search-16"
      >
        <div
          class="Connect-TranscriberBar--searchIcon-17"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-72"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
        <div
          class="MuiInputBase-root-143 Connect-TranscriberBar--inputRoot-18"
        >
          <input
            class="MuiInputBase-input-153 Connect-TranscriberBar--inputInput-19"
            placeholder="Searchâ€¦"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiAvatar-root-160 MuiAvatar-colorDefault-161 Connect-TranscriberBar--avatar-21"
      />
    </div>
  </header>
  <div
    class="WithData-Connect-ProjectTable---container-2"
  >
    <div
      class="MuiPaper-root-32 MuiPaper-elevation2-36 MuiPaper-rounded-33 WithData-Connect-ProjectTable---paper-3"
      id="ProjectTable"
    >
      <div
        class="WithData-Connect-ProjectTable---dialogHeader-5"
      >
        <div
          class="WithData-Connect-ProjectTable---grow-4"
        />
        <h5
          class="MuiTypography-root-107 MuiTypography-h5-123"
        >
          Choose Project
        </h5>
        <div
          class="WithData-Connect-ProjectTable---grow-4"
        />
        <button
          aria-label="Add"
          class="MuiButtonBase-root-69 MuiFab-root-163 MuiFab-primary-165 WithData-Connect-ProjectTable---button-9"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiFab-label-164"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-72 WithData-Connect-ProjectTable---icon-10"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-245"
          />
        </button>
      </div>
      <div
        class="RootBase-root-173"
      >
        <div
          class="MuiToolbar-root-59 MuiToolbar-regular-61 MuiToolbar-gutters-60 Toolbar-toolbar-174"
        >
          <div
            style="flex: 0 0 0px; margin-left: auto;"
          />
        </div>
        <div
          class="TableContainer-root-175"
        >
          <table
            class="MuiTable-root-180 Table-table-176"
            style="min-width: 800px;"
          >
            <colgroup>
              <col
                style="width: 200px;"
              />
              <col
                style="width: 400px;"
              />
              <col
                style="width: 100px;"
              />
              <col
                style="width: 100px;"
              />
              <col />
            </colgroup>
            <thead
              class="MuiTableHead-root-181"
            >
              <tr
                class="MuiTableRow-root-182 MuiTableRow-head-185"
              >
                <th
                  class="MuiTableCell-root-199 MuiTableCell-head-200 TableHeaderCell-cell-188 TableHeaderCell-cellNoWrap-198"
                  colspan="1"
                  scope="col"
                >
                  <div
                    class="TableHeaderCell-container-197"
                  >
                    <div
                      class="Content-content-211"
                    >
                      <div
                        class="SortLabel-root-214"
                      >
                        <span
                          class="MuiButtonBase-root-69 MuiTableSortLabel-root-227 SortLabel-sortLabelRoot-216 MuiTableSortLabel-active-228 SortLabel-sortLabelActive-218"
                          role="button"
                          tabindex="0"
                          title="Sort"
                        >
                          <span
                            class="Title-title-232"
                          >
                            Name
                          </span>
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root-72 MuiTableSortLabel-icon-229 MuiTableSortLabel-iconDirectionAsc-231"
                            focusable="false"
                            role="presentation"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                            />
                          </svg>
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ResizingControl-resizeHandle-233 TableHeaderCell-resizeHandle-189"
                  >
                    <div
                      class="TableHeaderCell-resizeHandleLine-190 ResizingControl-resizeHandleLine-234 ResizingControl-resizeHandleFirstLine-235"
                    />
                    <div
                      class="TableHeaderCell-resizeHandleLine-190 ResizingControl-resizeHandleLine-234 ResizingControl-resizeHandleSecondLine-236"
                    />
                  </div>
                </th>
                <th
                  class="MuiTableCell-root-199 MuiTableCell-head-200 TableHeaderCell-cell-188 TableHeaderCell-cellNoWrap-198"
                  colspan="1"
                  scope="col"
                >
                  <div
                    class="TableHeaderCell-container-197"
                  >
                    <div
                      class="Content-content-211"
                    >
                      <div
                        class="SortLabel-root-214"
                      >
                        <span
                          class="MuiButtonBase-root-69 MuiTableSortLabel-root-227 SortLabel-sortLabelRoot-216"
                          role="button"
                          tabindex="0"
                          title="Sort"
                        >
                          <span
                            class="Title-title-232"
                          >
                            Description
                          </span>
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root-72 MuiTableSortLabel-icon-229 MuiTableSortLabel-iconDirectionDesc-230"
                            focusable="false"
                            role="presentation"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                            />
                          </svg>
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ResizingControl-resizeHandle-233 TableHeaderCell-resizeHandle-189"
                  >
                    <div
                      class="TableHeaderCell-resizeHandleLine-190 ResizingControl-resizeHandleLine-234 ResizingControl-resizeHandleFirstLine-235"
                    />
                    <div
                      class="TableHeaderCell-resizeHandleLine-190 ResizingControl-resizeHandleLine-234 ResizingControl-resizeHandleSecondLine-236"
                    />
                  </div>
                </th>
                <th
                  class="MuiTableCell-root-199 MuiTableCell-head-200 TableHeaderCell-cell-188 TableHeaderCell-cellNoWrap-198"
                  colspan="1"
                  scope="col"
                >
                  <div
                    class="TableHeaderCell-container-197"
                  >
                    <div
                      class="Content-content-211"
                    >
                      <div
                        class="SortLabel-root-214"
                      >
                        <span
                          class="MuiButtonBase-root-69 MuiTableSortLabel-root-227 SortLabel-sortLabelRoot-216"
                          role="button"
                          tabindex="0"
                          title="Sort"
                        >
                          <span
                            class="Title-title-232"
                          >
                            Language
                          </span>
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root-72 MuiTableSortLabel-icon-229 MuiTableSortLabel-iconDirectionDesc-230"
                            focusable="false"
                            role="presentation"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                            />
                          </svg>
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ResizingControl-resizeHandle-233 TableHeaderCell-resizeHandle-189"
                  >
                    <div
                      class="TableHeaderCell-resizeHandleLine-190 ResizingControl-resizeHandleLine-234 ResizingControl-resizeHandleFirstLine-235"
                    />
                    <div
                      class="TableHeaderCell-resizeHandleLine-190 ResizingControl-resizeHandleLine-234 ResizingControl-resizeHandleSecondLine-236"
                    />
                  </div>
                </th>
                <th
                  class="MuiTableCell-root-199 MuiTableCell-head-200 TableHeaderCell-cell-188 TableHeaderCell-cellNoWrap-198"
                  colspan="1"
                  scope="col"
                >
                  <div
                    class="TableHeaderCell-container-197"
                  >
                    <div
                      class="Content-content-211"
                    >
                      <div
                        class="SortLabel-root-214"
                      >
                        <span
                          class="MuiButtonBase-root-69 MuiTableSortLabel-root-227 SortLabel-sortLabelRoot-216"
                          role="button"
                          tabindex="0"
                          title="Sort"
                        >
                          <span
                            class="Title-title-232"
                          >
                            Delete
                          </span>
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root-72 MuiTableSortLabel-icon-229 MuiTableSortLabel-iconDirectionDesc-230"
                            focusable="false"
                            role="presentation"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                            />
                          </svg>
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ResizingControl-resizeHandle-233 TableHeaderCell-resizeHandle-189"
                  >
                    <div
                      class="TableHeaderCell-resizeHandleLine-190 ResizingControl-resizeHandleLine-234 ResizingControl-resizeHandleFirstLine-235"
                    />
                    <div
                      class="TableHeaderCell-resizeHandleLine-190 ResizingControl-resizeHandleLine-234 ResizingControl-resizeHandleSecondLine-236"
                    />
                  </div>
                </th>
                <th
                  class="MuiTableCell-root-199 MuiTableCell-head-200 TableStubCell-cell-239"
                  colspan="1"
                  scope="col"
                />
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root-241"
            >
              <tr
                class="MuiTableRow-root-182"
              >
                <td
                  class="MuiTableCell-root-199 MuiTableCell-body-201 TableNoDataCell-cell-242"
                  colspan="5"
                >
                  <div
                    class="TableNoDataCell-textContainer-243"
                  >
                    <big
                      class="TableNoDataCell-text-244"
                    >
                      No data
                    </big>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
